### DISCUSSION TITLE
MLPoisson with free-space boundary conditions

### DISCUSSION THREAD

**Discussion by BenWibking** [AMReX-Codes/amrex#2419] [Q&A] [answered]
Is it possible to do a cell-centered Poisson solve with MLMG but with a FFT solve on level 0? This would be very useful for self-gravitating simulations with non-periodic boundary conditions using the zero-padding method ~~or, e.g., [this method](https://arxiv.org/abs/2103.08531).~~

The only other method I am aware of to do free-space boundary conditions for general mass distributions appears to be FMM-type methods, such as [McCorquodale05](https://crd.lbl.gov/assets/pubs_presos/AMCS/ANAG/HPSEC05.pdf), but this appears to be significantly more complicated but may require less communication.

**Comment by zingale** [comment, marked as answer]
Ben, you should look at the method on Castro which uses a multipole method to get Dirichlet BCs for the multigrid solve.  It works well.  See Katz et al. 2016 for details.

**Reply by BenWibking** [reply]
I looked at that, but I don't think that will work for my applications.  There is some discussion in [Moon et al. (2019)](https://ui.adsabs.harvard.edu/abs/2019ApJS..241...24M/abstract) about such an multipole expansion not working well for thin disks that extend close to the domain boundaries ("Although this method appears feasible for small lmax and mmax, the computational cost would increase to ${ \mathcal O }({l}_{\max }{m}_{\max }{N}^{4})$ for a flattened mass distribution with surfaces lying close to domain boundaries.")

In another case of interest, I don't think this will work for star clusters that may blow out half of their mass through the open boundaries. But if there's an example similar to either of these that works with the multipole expansion method, I'd be happy to take a look.

Alternatively, I would probably go the route of the [James (1977)](https://ui.adsabs.harvard.edu/abs/1977JCoPh..25...71J/abstract) method.

**Reply by maximumcats** [reply]
I agree that the multipole expansion is not a great idea if you will ever have significant mass near the boundaries. In that cames James would be better. We tried it several years ago and it's relatively straightforward to implement but I never got the convergence properties right. The paper by Sanghyuk Moon and others does look promising and likely addresses the reason it didn't work well in our implementation, so I agree it's worth exploring, although I think there are some subtleties in any calculation of the surface charge term.  If there was a motivating use case I'd be happy to investigate implementing it again.

**Reply by BenWibking** [reply]
The motivating use case for me is simulating star clusters. ORION has used a similar multipole method and just enlarged the domain so that no significant mass would ever be near the boundaries. However with Quokka, I'm concerned with increasing the number of levels on GPU while still trying to get reasonable strong scaling for simulations with already a very large dynamic range (say, 8 levels).  Also, for timescales significantly longer than the global free-fall timescale of the cloud (possibly de-refining in order to make this feasible), mass will eventually leave the box, unless it's allowed to dynamically expand (which could be difficult to deal with).

If the free-space gravity solver could be shared between Quokka and Castro, I'd be happy to collaborate on this as well, although I'm using AmrCore directly rather than AmrLevel.

**Reply by WeiqunZhang** [reply]
I don't think it's too hard to implement the method in McCorquodale et al., which followed James.  (1) Solve the Poisson equation on the original grids using Dirichlet BC.  We already have a multigrid solver for this.  (2) Compute surface charge density on the domain face.  Grow the domain and use the multipole method to compute the potential on the grown domain's face from the surface charge.  We can probably reuse some of Max's work from Castro.  (3) Solve the Poisson equation on the grown domain using Dirichlet BC.  This step is straightforward.

I am interested in this because WarpX would also like to have such a solver.

**Reply by BenWibking** [reply]
Ah, I see. I didn't recognize the connection between the methods. It looks like the only difference is how the potential due to the surface charge is computed. In the node-centered case, the surface charge _seems_ straightforward, but I am still not totally clear on how to compute the surface charge when the density/potential are cell-centered. @maxpkatz Is this the subtlety you were referring to?

I think it would be ideal if this method were incorporated into AMReX, or could be made into an example code/miniapp that all three of our codes could use as a template.

**Reply by BenWibking** [reply]
Also- it's not clear what multipoles McCorquodale et al. use, but perhaps (2D) Cartesian multipoles might make sense for this? I'm not sure that spherical harmonics make sense for 2D patches of charge.

**Reply by BenWibking** [reply]
@WeiqunZhang is this a best way to use AMReX communication primitives to do the all-to-all exchange of multipoles? I'm not entirely sure what is the best parallel algorithm for summing the multipoles on all of the exterior faces.

Also, what does WarpX currently do for its Poisson solve?

**Reply by WeiqunZhang** [reply]
Re: mutipole, I don't know what's the best approach for this.  I guess we will have to try.  If you can work on it, I will work on (1), (3) and the first part of (2).

WarpX's Poisson solver assumes the BC is homogeneous Dirichlet.  Because there are both positive and negative charges, it's probably a better approximation than the case of gravity.

**Reply by BenWibking** [reply]
I'm currently pretty busy with things that have deadlines in the next few weeks, but I should have time to work on the implementation after that.

Just for reference, I found this [tech report](https://www.osti.gov/servlets/purl/929486) with more details on the McCorquodale+ method. They do in fact use 2D cartesian multipoles for the potential of the induced surface charge. However, everything in that work is node-centered. I'm not sure whether WarpX uses cell-centered or node-centered. For my code at least, I will need to work out the changes in stencil etc., for cell-centered potential/charge.

**Reply by WeiqunZhang** [reply]
Maybe I am missing something.  I don't see the index type is an issue.  WarpX's solver is node based.  We can support both.

**Reply by BenWibking** [reply]
Yes, either case should work. I just meant that I think the calculation of the surface charge is slightly different between the two cases (although I haven't checked yet to be sure, it might actually be the same).

**Reply by WeiqunZhang** [reply]
@BenWibking Could you give this a try? https://github.com/AMReX-Codes/amrex/pull/2912

**Reply by BenWibking** [reply]
This will be great to have. Unfortunately, I'm in the middle of moving back to the US from Australia at the moment, so I won't be able to try it out with Quokka for at least a few weeks.
